service/history/api/get_history_util.go

```diff
@@ -93,7 +93,13 @@ func GetRawHistory(
 	metricsHandler := interceptor.GetMetricsHandlerFromContext(ctx, shardContext.GetLogger()).WithTags(metrics.OperationTag(metrics.HistoryGetHistoryScope))
 	metrics.HistorySize.With(metricsHandler).Record(int64(size))
 
-	if len(nextToken) == 0 && transientWorkflowTaskInfo != nil {
+	// If
+	//    there are no more events in DB (i.e., page token is empty),
+	//    and transient/speculative WFT events are present (i.e., such WFT is in MS),
+	//    and WF is still running (i.e., last event is not completion event),
```

Comment:
this is for handling the case where transient/speculative workflow task is only scheduled so they are still in ms even when workflow is closed (terminated)?

Reply:
yes, but not only. History is loaded outside of the WF lock, and during testing I got a case when race between normal WF completion and polling for WF result returned speculative WFT events which were there before completion, but got attached after `WorkflowCompleted` event.